app/models/onboarding_form.rb

```diff
@@ -62,50 +57,101 @@ class OnboardingForm
+  validates :parts, :presence => true
+
+  validates_with EnvironmentValidator, if: Proc.new{|form| form.part_submitted? 'environment'}
+  validates_with ServiceValidator, if: Proc.new{|form| form.part_submitted? 'service'}
+  validates_with MSAValidator, if: Proc.new{|form| form.part_submitted? 'msa3'}
+  validates_with IntegrationValidator, if: Proc.new{|form| form.part_submitted? 'integration'}
+  validates_with ContactValidator, if: Proc.new{|form| form.part_submitted? 'contact'}
 
   def initialize(attributes)
     super
   end
 
+  def part_submitted?(part)
+    parts.split(',').include? part
+  end
+
+  def is_integration?
+    environment_access == ENVIRONMENT_ACCESS_INTEGRATION
+  end
+
+  def is_vsp?
+    service_provider == 'VSP'
+  end
+
+  def is_reuse_service_config?
+    reuse_service_config == 'Yes'
+  end
+
+  def is_msa?
+    matching_service_adapter == 'MSA'
+  end
+
+  def is_reuse_msa_config?
+    reuse_msa_config == 'Yes'
+  end
+
   def validate_entity_ids_are_different
     if (@service_entity_id == @matching_service_entity_id)
       errors['matching_service_entity_id'] << 'needs to be different from the Verify service entity ID'
     end
   end
 
-  def validate_certificate
-    CERTIFICATE_FIELDS.each {|certificate_field|
-      begin
-        certificate = OpenSSL::X509::Certificate.new(self.send(certificate_field))
-        if is_integration_access_form?
-          validate_certificate_issuer(certificate, CERTIFICATE_ISSUER_INTEGRATION, certificate_field, 'integration')
-        else
-          validate_certificate_issuer(certificate, CERTIFICATE_ISSUER_PRODUCTION, certificate_field, 'production')
-        end
-      rescue
-        errors[certificate_field] << 'is malformed'
-      end
-    }
+  def validate_not_empty(field_name)
+    if self.send(field_name).empty?
+      errors.add(field_name, 'must not be empty')
+    end
   end
 
-  def validate_certificate_issuer(certificate, issuer, certificate_field, environment)
-    if !certificate.issuer.to_s.include?("CN=#{issuer}")
-      errors.add(certificate_field, "Certificates for the #{environment} environment must be issued by #{issuer}")
+  def validate_url(field_name)
+    unless self.send(field_name)[/\Ahttps?:\/\//]
+      errors.add(field_name, 'must be a url')
     end
   end
 
-  def is_integration_access_form?
-    environment_access == ENVIRONMENT_ACCESS_INTEGRATION
+  def validate_email(field_name)
+    unless self.send(field_name)[/.+@.+/]
+      errors.add(field_name, 'is not properly formatted')
+    end
+  end
+
+  def validate_strong_password(field_name)
+    if self.send(field_name).length < 8
+      errors.add(field_name, 'must be at least 8 characters')
+    end
+  end
+
+  def validate_certificate(field_name)
+    begin
+      certificate = convert_value_to_x509_certificate(self.send(field_name))
+    rescue
+      errors[field_name] << "is malformed"
+    end
+    validate_certificate_issuer(certificate, field_name)
+  end
```

Comment:
Could use ensure here (don't need the begin/end):

```
def validate_certificate(field_name)
  certificate = convert_value_to_x509_certificate(self.send(field_name))
rescue
  errors[field_name] << "is malformed"
ensure
  validate_certificate_issuer(certificate, field_name)
end
```

Though if the `OpenSSL::X509::Certificate.new(Base64.decode64(cert_value))` errors might be good to do `validate_certificate_issuer(certificate, field_name) unless certificate.nil?`

Reply:
I reworked this here in a different way.